<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.14"/>
<meta name="viewport" content="width=device-width, initial-scale=1"/>
<title>My Project: trimFilterDS user manual</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/searchdata.js"></script>
<script type="text/javascript" src="search/search.js"></script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td id="projectalign" style="padding-left: 0.5em;">
   <div id="projectname">My Project
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.14 -->
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
var searchBox = new SearchBox("searchBox", "search",false,'Search');
/* @license-end */
</script>
<script type="text/javascript" src="menudata.js"></script>
<script type="text/javascript" src="menu.js"></script>
<script type="text/javascript">
/* @license magnet:?xt=urn:btih:cf05388f2679ee054f2beb29a391d25f4e673ac3&amp;dn=gpl-2.0.txt GPL-v2 */
$(function() {
  initMenu('',true,false,'search.php','Search');
  $(document).ready(function() { init_search(); });
});
/* @license-end */</script>
<div id="main-nav"></div>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
</div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

</div><!-- top -->
<div class="header">
  <div class="headertitle">
<div class="title">trimFilterDS user manual </div>  </div>
</div><!--header-->
<div class="contents">
<div class="textblock"><p>This program reads two paired end <code>fastq</code> files as an input and filters them according to the following criteria:</p><ul>
<li>Discard/trims reads containing adapter remnants.</li>
<li>Discards reads matching contaminations (sequences collected in a <code>fasta</code> file or in an <code>idx</code> file created by <code>makeTree</code>, <code>makeBloom</code></li>
<li>Discards/trims low quality reads.</li>
<li>Discards/trims reads containing N base callings.</li>
</ul>
<p>If one of the two rads is discarded, the corresponding paired read is automatically discarded.</p>
<h2>Running the program</h2>
<p>Usage <code>C</code> executable (in folder <code>bin</code>):</p>
<div class="fragment"><div class="line">Usage: trimFilterDS --ifq &lt;INPUT1.fq&gt;:&lt;INPUT2.fq&gt; --length &lt;READ_LENGTH&gt;</div><div class="line">                  --output [O_PREFIX] --gzip [y|n]</div><div class="line">                  --adapter [&lt;AD1.fa&gt;:&lt;AD2.fa&gt;:&lt;mismatches&gt;:&lt;score&gt;]</div><div class="line">                  --method [TREE|BLOOM]</div><div class="line">                  (--idx [&lt;INDEX_FILE&gt;:&lt;score&gt;:&lt;lmer_len&gt;] |</div><div class="line">                   --ifa [&lt;INPUT.fa&gt;:&lt;score&gt;:[lmer_len]])</div><div class="line">                  --trimQ [NO|ALL|ENDS|FRAC|ENDSFRAC|GLOBAL]</div><div class="line">                  --minL [MINL]  --minQ [MINQ]</div><div class="line">                  (--percent [percent] | --global [n1:n2])</div><div class="line">                  --trimN [NO|ALL|ENDS|STRIP]</div><div class="line">Reads in paired end fq files (gz, bz2, z formats also accepted) and removes:</div><div class="line">  * low quality reads,</div><div class="line">  * reads containing N base callings,</div><div class="line">  * reads representing contaminations, belonging to sequences in INPUT.fa</div><div class="line">Outputs 4 [O_PREFIX]_fq.gz files per input fastq file containing: &quot;good&quot; reads,</div><div class="line">discarded low Q reads discarded reads containing N&#39;s, discarded contaminations.</div><div class="line"></div><div class="line">If one read is removed, its corresponding paired read is removed as well.</div><div class="line"></div><div class="line">Options:</div><div class="line"> -v, --version prints package version.</div><div class="line"> -h, --help    prints help dialog.</div><div class="line"> -f, --ifq     2 fastq input files [*fq|*fq.gz|*fq.bz2] separated by colons, mandatory option.</div><div class="line"> -l, --length  read length: length of the reads, mandatory option.</div><div class="line"> -o, --output  output prefix (with path), optional (default ./out). Output</div><div class="line"> -z, --gzip    gzips output files: yes or no (default yes).</div><div class="line"> -A, --adapter adapter input four fields separated by colons:</div><div class="line">               &lt;AD1.fa&gt;: fasta file containing adapters from read 1,</div><div class="line">               &lt;AD2.fa&gt;: fasta file containing adapters from read 2,</div><div class="line">               &lt;mismatches&gt;: maximum mismatch count allowed,</div><div class="line">               &lt;score&gt;: score threshold  for the aligner.</div><div class="line"> -x, --idx     index input file. To be included with any method. 3 fields</div><div class="line">               3 fields separated by colons:</div><div class="line">               &lt;INDEX_FILE&gt;: output of makeTree, makeBloom,</div><div class="line">               &lt;score&gt;: score threshold to accept a match [0,1],</div><div class="line">               [lmer_len]: correspond to the length of the lmers to be</div><div class="line">                        looked for in the reads [1,READ_LENGTH].</div><div class="line"> -a, --ifa     fasta input file. To be included only with method TREE</div><div class="line">               (it excludes the option --idx). Otherwise, an</div><div class="line">               index file has to be precomputed and given as parameter</div><div class="line">               (see option --idx). 3 fields separated by colons:</div><div class="line">               &lt;INPUT.fa&gt;: fasta input file [*fa|*fa.gz|*fa.bz2],</div><div class="line">               &lt;score&gt;: score threshold to accept a match [0,1],</div><div class="line">               &lt;lmer_len&gt;: depth of the tree: [1,READ_LENGTH]. It will</div><div class="line">                        correspond to the length of the lmers to be</div><div class="line">                        looked for in the reads.</div><div class="line"> -C, --method  method used to look for contaminations:</div><div class="line">               TREE:  uses a 4-ary tree. Index file optional,</div><div class="line">               BLOOM: uses a bloom filter. Index file mandatory.</div><div class="line"> -Q, --trimQ   NO:       does nothing to low quality reads (default),</div><div class="line">               ALL:      removes all reads containing at least one low</div><div class="line">                         quality nucleotide.</div><div class="line">               ENDS:     trims the ends of the read if their quality is</div><div class="line">                         below the threshold -q,</div><div class="line">               FRAC:     discards a read if the fraction of bases whose</div><div class="line">                         quality lies below</div><div class="line">                         the threshold -q is over 5 percent or a user</div><div class="line">                         defined percentage in -p.</div><div class="line">               ENDSFRAC: trims the ends and then discards the read if</div><div class="line">                         there are more low quality nucleotides than the</div><div class="line">                         allowed by the option -p.</div><div class="line">               GLOBAL:   removes n1 cycles on the left and n2 on the</div><div class="line">                         right, specified in -g.</div><div class="line">               All reads are discarded if they are shorter than MINL.</div><div class="line"> -m, --minL    minimum length allowed for a read before it is discarded</div><div class="line">               (default 25).</div><div class="line"> -q, --minQ    minimum quality allowed (int), optional (default 27).</div><div class="line"> -p, --percent percentage of low quality bases to be admitted before</div><div class="line">               discarding a read (default 5),</div><div class="line"> -g, --global  required option if --trimQ GLOBAL is passed. Two int,</div><div class="line">               n1:n2, have to be passed specifying the number of cycles</div><div class="line">               to be globally cut from the left and right, respectively.</div><div class="line"> -N, --trimN   NO:     does nothing to reads containing N&#39;s,</div><div class="line">               ALL:    removes all reads containing N&#39;s,</div><div class="line">               ENDS:   trims ends of reads with N&#39;s,</div><div class="line">               STRIPS: looks for the largest substring with no N&#39;s.</div><div class="line">               All reads are discarded if they are shorter than minL.</div></div><!-- fragment --><p>NOTE: the parameters -l or &ndash;length are meant to identify the length of the reads in the input data. Actually, <code>trimFilterDS</code> also copes with data holding reads with different lengths. The length parameter must hold the length of the longest read in the dataset.</p>
<h2>Output description</h2>
<ul>
<li><code>[O_PREFIX1 | O_PREFIX2]_good.fq.gz</code>: contains reads that passed all filters (maybe trimmed).</li>
<li><code>[O_PREFIX1 | O_PREFIX2]_adap.fq.gz</code>: contains reads discarded due to the presence of adapters.</li>
<li><code>[O_PREFIX1 | O_PREFIX2]_cont.fq.gz</code>: contains contamination reads.</li>
<li><code>[O_PREFIX1 | O_PREFIX2]_lowQ.fq.gz</code>: contains reads discarded due to low quality issues.</li>
<li><code>[O_PREFIX1 | O_PREFIX2]_NNNN.fq.gz</code>: contains reads discarded due to <em>N</em>'s issues.</li>
<li><code>[O_PREFIX1 | O_PREFIX2]_summary.bin</code>: binary file where information about the filtering process is stored. Structure of the file.<ul>
<li>filters, <code>4*sizeof(int) Bytes</code>: array of int with entries <code>i = {ADAP(0), CONT(1), LOWQ(2), NNNN(3)}</code>. A given entry takes the value of the filter it was applied to and 0 otherwise. <code>filters[ADAPT] = {0,1}</code>, <code>filters[CONT] = {NO(0), TREE(1), BLOOM(2)}</code>, <code>filters[LOWQ] = {NO(0), ALL(1), ENDS(2), FRAC(3), ENDSFRAC(4), GLOBAL(5)}</code>, <code>filters[trimN] = {NO(0), ALL(1), ENDS(2), STRIPS(2)}</code>.</li>
<li>trimmed, <code>4*sizeof(int) Bytes</code>: array of integers with entries i = {ADAP(0), CONT(1), LOWQ(2), NNNN(3)}, containing how many reads were trimmed due to the corresponding filter.</li>
<li>discarded, <code>4*sizeof(int) Bytes</code>: array of integers with entries i = {ADAP(0), CONT(1), LOWQ(2), NNNN(3)}, containing how many reads were discarded due to the corresponding filter.</li>
<li>good, <code>sizeof(int) Bytes</code>: number of accepted reads (maybe trimmed).</li>
<li>nreads, <code>sizeof(int) Bytes</code>: total number of reads.</li>
</ul>
</li>
</ul>
<h2>Filters</h2>
<h4>Adapters</h4>
<p>Technical sequences within the reads are detected if the option <code>--adapters &lt;ADAPTERS1.fa&gt;:&lt;ADAPTERS2.fa&gt;:&lt;mismatches&gt;:&lt;score&gt;</code> is given. The adapter(s) sequence(s) are read from the fasta files, and then prepended to their respective reads. Then, a 'seed and extend' approach is used to look for overlaps following the same rules followed in the single end case. See <code>README_trimFilter.md</code> for details on how two matching subsequences are detected and how the score is computed. The paired reads are correspondingly trimmed, removed or left as is. The following figure describes possible cases:</p>
<div class="image">
<img src="./pics/adapters/palindrome_new.png" alt="noimage" title="Adapters identification"/>
</div>
 <h4>Impurities</h4>
<p>Contaminations are removed if a fasta file or an index file are given as an input. The methods provided to look for contaminations work in the very same way as they work for single end data. If one of the reads is discarded, then, the other read is discarded as well. See <code>README_trimFilter.md</code> for more details on how the contaminations are handled.</p>
<h4>LowQ</h4>
<p>Again, the detection and trimming/removal of reads containing low quality nucleotides is done following the same procedure as for single end data. We list the options below, see <code>README_trimFilter.md</code> for more details.</p>
<ul>
<li><code>--trimQ NO</code> or flag absent: nothing is done to the reads with low quality.</li>
<li><code>--trimQ ALL</code>: all reads containing at least one low quality nucleotide are redirected to <code>*_lowq.fq.gz</code></li>
<li><code>--trimQ ENDS</code>: look for low quality (below MINQ) base callings at the beginning and at the end of the read. Trim them at both ends until the quality is above the threshold. Keep the read in <code>*_good.fq.gz</code> and annotate in the fourth line where the read has been trimmed (starting to count from 0) if the length of the remaining part is larger than <code>MINL</code>. Redirect the read to <code>*_lowq.fq.gz</code> otherwise.</li>
<li><code>--trim FRAC [--percent p]</code>: redirect the read to <code>*_lowq.fq.gz</code> if there are more than <code>p%</code> nucleotides whose quality lies below the threshold. <code>p=5</code> per default.</li>
<li><code>--trim ENDSFRAC --percent p</code>: first trim the ends as in the <code>ENDS</code> option. Accept the trimmed read if the number of low quality nucleotides does not exceed <code>p%</code> (default <code>p = 5</code>). Redirect the read to <code>*_lowq.fq.gz</code> otherwise.</li>
<li><code>--trim GLOBAL --global n1:n2</code>: cut all read globally <code>n1</code> nucleotides from the left and <code>n2</code> from the right.</li>
</ul>
<p><b>Note:</b> qualities are evaluated assuming the reads to follow the L - Illumina 1.8+ Phred+33, convention, see <a href="https://en.wikipedia.org/wiki/FASTQ_format#Encoding">Wikipedia</a>. Adjust the values for a different convention.</p>
<h4>N trimming</h4>
<p>We allow for the following options (see README_trimFilter.md for examples and more details):</p>
<ul>
<li><code>--trimN NO</code> (or flag absent): Nothing is done to the reads containing N's.</li>
<li><code>--trimN ALL</code>: All reads containing at least one N are redirected to <code>*_NNNN.fq.gz</code></li>
<li><code>--trimN ENDS</code>: N's are trimmed if found at the ends, left "as is" otherwise. If the trimmed read length is smaller than MINL, it is discarded.</li>
<li><code>--trimN STRIP</code>: Obtain the largest N free subsequence of the read. Accept it if is longer than the half of the original read length, redirect it to <code>*_NNNN.fq.gz</code> otherwise.</li>
</ul>
<h2>Test/examples</h2>
<p>The examples in folder <code>examples/trimFilterDS_SReport/</code> work in the following way:</p><ol type="1">
<li>See folder <code>fa_fq_files</code>. The files <code>EColi_rRNA_DS.read1.fq.gz</code> and <code>EColi_rRNA_DS.read2.fq.gz</code> were created with <code>create_fq.sh</code> and contain:<ul>
<li>1000 reads of length 50 from <code>EColi_genome.fa</code> with NO errors.</li>
<li>5000 reads of length 50 from <code>rRNA_modified.fa</code> with NO errrors (rRNA contaminations).</li>
<li>Artificially generated reads with low quality score (see <code>create_fq.sh</code>)</li>
<li>Artificially generated reads with Ns (see <code>create_fq.sh</code>).</li>
</ul>
</li>
<li><code>run_example_TREE.sh</code>: the code was tested with flags: `../../bin/trimFilter -l 50 &ndash;ifq\ &ndash;ifq ../fa_fq_files/EColi_rRNA_DS.read1.fq.gz:../fa_fq_files/EColi_rRNA_DS.read1.fq.gz &ndash;method TREE &ndash;ifa ../fa_fq_files/rRNA_modified.fa:0.2:30 \ &ndash;trimQ ENDSFRAC &ndash;trimN ENDS -o treeDS &ndash;adapters \ ../fa_fq_files/ad_read1.fa:../fa_fq_files/ad_read2.fa:2:40 i.e., we check for contaminations from rRNA, trim reads with lowQ at the ends and less than 5% in the remaining part, and strip reads containing N's at the ends.</li>
<li><code>run_example_BLOOM.sh</code>: trimFilterDS is run like in 2. but passing a bloom filter to look for contaminations with <code>score=0.4</code> and the &ndash;trimN STRIP option.</li>
<li>With this set up, it is possible to run further customized tests.</li>
<li>See the folder <code>adapters</code> for examples on adapter contaminations (and its corresponding README file).</li>
</ol>
<p>NOTE: <code>rRNA_modified.fa</code> is the <code>rRNA_CRUnit.fa</code> sequence, where we have removed the lines containing N's for testing purposes.</p>
<h2>Contributors</h2>
<p>Paula PÃ©rez Rubio</p>
<h2>License</h2>
<p>GPL v3 (see LICENSE.txt) </p>
</div></div><!-- contents -->
<!-- start footer part -->
<hr class="footer"/><address class="footer"><small>
Generated by &#160;<a href="http://www.doxygen.org/index.html">
<img class="footer" src="doxygen.png" alt="doxygen"/>
</a> 1.8.14
</small></address>
</body>
</html>
